### PR TITLE
fix(efb): metar parser error

### PR DESCRIPTION
**Pull request by salvapatuel**
## Summary of Changes
Fixes an error in the METAR Parser
The latest METAR parser throws an error when the ICAO does not have a METAR available

## Screenshots (if necessary)
Error: 
image

Fix:
image

## Additional context
Simple Error handling added

## Testing instructions
Use an ICAO code that does not have a METAR station (ex: SAVE)

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Review by Benjozork — APPROVED** [review]

**Comment by frankkopp** [comment]
Thanks for this fix.
But could you please upload your images to github (into the comment) and not use some obscure links?

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,7 @@
 
 ## 0.8.0
 
+1. [EFB] Fix Metar parser error - @brainshot (brainshot)
```
Comment (frankkopp) [inline .github/CHANGELOG.md:9]: Please move this to the end of the list.

**Review by frankkopp — CHANGES_REQUESTED** [review]

File: .github/CHANGELOG.md (hunk 1 above)
Comment (salvapatuel) [reply .github/CHANGELOG.md]: done

**Comment by frankkopp** [comment]
Thanks!

We should add a unit test for this:
https://github.com/flybywiresim/a32nx/blob/master/src/instruments/src/EFB/Utils/parseMetar.spec.ts

If you know how, you can do it otherwise I will do it when I have the time.

**Comment by frankkopp** [comment]
I have found additional issue when using different METAR sources. I will refactor the reading ant then parsing of metar a bit to get better error messages.

**Comment by frankkopp** [comment]
@salvapatuel 
I have added unit tests and refactored the weather widget to better distinguish between retrieving errors from parsing errors. 
Hope you are ok with these changes as well - maybe you can double check. 

I have noticed that PilotEdge as METAR source does not return any METAR at all any more. @Benjozork do you now if something changed for PilotEdge?

**Review by frankkopp — APPROVED** [review]

**Comment by mico975** [comment]
Could someone please add some testing instructions on when which error message should pop up and possibly how to provoke other messages? I only manage to get NO VALID ICAO CHOSEN and no other message while multiple are clearly included in the code.

**Comment by salvapatuel** [comment]
> Could someone please add some testing instructions on when which error message should pop up and possibly how to provoke other messages? I only manage to get NO VALID ICAO CHOSEN and no other message while multiple are clearly included in the code.

I think there should be 2 Error messages, one for "INVALID ICAO CODE" and the other one "METAR NOT AVAILABLE" , as the later the ICAO code is correct but there is no station available on the selected source.

**Comment by ShadowFrost3** [comment]
Quality Assurance Trainee Report


Object of testing: #6880
Tier of Testing:1
Date: 16/03/2022

Testing Process: Put a bunch of letters in the METAR box and see what happens 
Negatives: 
Testing Results: Passed

Conclusions: the 2nd box (on the right) sometimes took a few seconds to catch up. I don't know why but its not a huge concern. Never occurred with the left box. Bias? Fallacy? Perception error? Who knows.. 

image

**Comment by mico975** [comment]
Quality Assurance Tester Report

Discord          : mico#3145
Object of testing: #6880
Tier of Testing  : 1
Date             : 17/03/2022

Testing Process: 
Requested various METARs. - OK

Negatives:       
Requesting METAR without populated Simbrief and then removing ICAO results in the NO VALID ICAO CHOSEN OR NO METAR FOR ICAO AVAILABLE error message, when the NO VALID ICAO CHOSEN would be more appropriate (see media below). Simbrief fallback with "empty" field works fine with ICAO in gray and METAR displayed.

Testing Results: 
Passed

Conclusions: 
Message in negatives is technically correct, but logic for it could be better.

Media:       
image
image

**Comment by frankkopp** [comment]
@ShadowFrost3 @mico975 

I have made three error cases now for the ICAO input itself:

- "NO ICAO PROVIDED" 
  - shown when input <4 characters
  - see [1]
- "ICAO INVALID OR NO METAR AVAILABLE" 
  - shown when the Metar API for the ICAO returns an error (METAR for ICAO has not been found)
  - see [1]
- "RECEIVED METAR COULD NOT BE PARSED..."
  - shown when the Metar API did not return an error but the returned METAR string is not valid
  - see [2] - (this is a fake error - the metar is ok but I wanted to show the err message)
  
[1] https://user-images.githubusercontent.com/16833201/158997223-a1e17467-5c34-4a03-91d9-b38c9156fd14.mp4
[2] https://user-images.githubusercontent.com/16833201/158997227-cef1ce73-7764-41df-bbf4-9e692d8cb744.mp4

Would this be acceptable for you guys?
Do you want to re-test?

**Pull request merged**